mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs

```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
+
+		internal override byte[] DynamicData () {
+			return sun_path;
+		}
+		internal override long GetDynamicLength () {
+			return sun_path_len;
+		}
+		internal override void SetDynamicLength (long value) {
+			sun_path_len = value;
+		}
+
+		[DllImport (Syscall.MPH, SetLastError=true,
+				EntryPoint="Mono_Posix_SockaddrUn_get_sizeof_sun_path")]
+		static extern int get_sizeof_sun_path ();
+		static readonly int sizeof_sun_path = get_sizeof_sun_path ();
+
+		public SockaddrUn () : base (SockaddrType.SockaddrUn | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNIX)
+		{
+			sun_path = new byte[sizeof_sun_path];
+			sun_path_len = 0;
+		}
+		public SockaddrUn (int size) : base (SockaddrType.SockaddrUn | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNIX)
+		{
+			sun_path = new byte[size];
+			sun_path_len = 0;
+		}
+		public SockaddrUn (string path, bool @abstract = false) : base (SockaddrType.SockaddrUn | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNIX)
+		{
+			var bytes = UnixEncoding.Instance.GetBytes (path);
+			if (@abstract) {
+				sun_path = new byte[1 + bytes.Length];
+				Array.Copy (bytes, 0, sun_path, 1, bytes.Length);
+			} else {
+				sun_path = bytes;
+			}
+			sun_path_len = sun_path.Length;
+		}
+
+		public bool IsAbstract {
+			get {
+				return sun_path_len > 0 && sun_path[0] == 0;
+			}
+		}
+		public string Path {
+			get {
+				var offset = IsAbstract ? 1 : 0;
+				// Remove data after null terminator
+				int length;
+				for (length = 0; offset + length < sun_path_len; length++)
+					if (sun_path[offset + length] == 0)
+						break;
+				return UnixEncoding.Instance.GetString (sun_path, offset, length);
+			}
+		}
+
+		public override string ToString ()
+		{
+			return string.Format ("{{sa_family={0}, sun_path=\"{1}{2}\"}}", sa_family, IsAbstract ? "\\0" : "", Path);
+		}
+
+		public static new SockaddrUn FromSockaddrStorage (SockaddrStorage storage)
+		{
+			// This will make the SockaddrUn larger than it needs to be (because
+			// storage.data_len includes the sun_family field), but it will be
+			// large enough.
+			var ret = new SockaddrUn ((int) storage.data_len);
+			storage.CopyTo (ret);
+			return ret;
+		}
+
+		public override int GetHashCode ()
+		{
+			return sun_family.GetHashCode () ^ IsAbstract.GetHashCode () ^ Path.GetHashCode ();
+		}
+		public override bool Equals (object obj)
+		{
+			if (!(obj is SockaddrUn))
+				return false;
+			return Equals ((SockaddrUn) obj);
+		}
+		public bool Equals (SockaddrUn value)
+		{
+			if (value == null)
+				return false;
+			return sun_family == value.sun_family
+				&& IsAbstract == value.IsAbstract
+				&& Path == value.Path;
+		}
+	}
+
+	[Map ("struct sockaddr_in")]
+	[CLSCompliant (false)]
+	[StructLayout (LayoutKind.Sequential)]
+	public class SockaddrIn : Sockaddr, IEquatable<SockaddrIn> {
+		public UnixAddressFamily sin_family { // AF_INET
+			get {
```

Comment:
Style: You can use a single-line form here:

    get { return sa_family; }
    set { sa_family = value; }